docs/source/02_installation/02_configuration.md
# Configuration

Plugin maintains the configuration in the `conf/base/airflow-k8s.yaml` file.

```yaml
# Base url of the Apache Airflow, should include the schema (http/https)
host: https://airflow.example.com

# Directory from where Apache Airflow is reading DAGs definitions
output: gs://airflow-bucket-example-com

# Configuration used to run the pipeline
run_config:

    # Name of the image to run as the pipeline steps
    image: airflow-k8s-plugin-demo

    # Pull policy to be used for the steps. Use Always if you push the images
    # on the same tag, or Never if you use only local images
    image_pull_policy: IfNotPresent
    
    # Pod startup timeout in seconds - if timeout passes the pipeline fails, default to 600 
    startup_time: 600

    # Namespace for Airflow pods to be created
    namespace: airflow

    # Name of the Airflow experiment to be created
    experiment_name: Airflow K8S Plugin Demo

    # Name of the dag as it's presented in Airflow
    run_name: airflow-k8s-plugin-demo

    # Apache Airflow cron expression for scheduled runs
    cron_expression: "@daily"
        
    # Optional start date in format YYYYMMDD, if not provided `days_ago(2)` is used instead
    start_date: "20210721"

    # Optional pipeline description
    description: "Very Important Pipeline"

    # Comma separated list of image pull secret names
    image_pull_secrets: my-registry-credentials
   
    # Service account name to execute nodes with
    service_account_name: airflow

    # List of handlers executed after task failure 
    failure_handlers:
      # type of integration, currently only slack is available
      - type: slack
        # airflow connection id with following parameters:
        # host - webhook url
        # password - webhook password
        # login - username
        connection_id: slack
        # message template that will be send. It can contains following parameters that will be replaced:
        # task
        # dag
        # execution_time
        message_template: |
            :red_circle: Task Failed.
            *Task*: {task}
            *Dag*: {dag}
            *Execution Time*: {execution_time}
            *Log Url*: {url}
            
    # Optional volume specification
    volume:
        # Storage class - use null (or no value) to use the default storage
        # class deployed on the Kubernetes cluster
        storageclass: # default
        # The size of the volume that is created. Applicable for some storage
        # classes
        size: 1Gi
        # Access mode of the volume used to exchange data. ReadWriteMany is
        # preferred, but it is not supported on some environements (like GKE)
        # Default value: ReadWriteOnce
        #access_modes: [ReadWriteMany]
        # Flag indicating if the data-volume-init step (copying raw data to the
        # fresh volume) should be skipped
        skip_init: False
        # Allows to specify fsGroup executing pipelines within containers
        # Default: root user group (to avoid issues with volumes in GKE)
        owner: 0
        # If set to True, shared persistent volume will not be created at all and all other parameters under
        # `volume` are discarded
        disabled: False

    # List of optional secrets specification
    secrets:
            # deploy_type: (Optional - default: 'env`) The type of secret deploy in Kubernetes, either `env` or `volume`
        -   deploy_type: "env"
            # deploy_target: (Optional) The environment variable when `deploy_type` `env` or file path when `deploy_type` `volume` where expose secret. If `key` is not provided deploy target should be None.
            deploy_target: "SQL_CONN"
            # secret: Name of the secrets object in Kubernetes
            secret: "airflow-secrets"
            # key: (Optional) Key of the secret within the Kubernetes Secret if not provided in `deploy_type` `env` it will mount all secrets in object
            key: "sql_alchemy_conn"


    # Apache Airflow macros to be exposed for the parameters
    # List of macros can be found here:
    # https://airflow.apache.org/docs/apache-airflow/stable/macros-ref.html
    macro_params: [ds, prev_ds]

    # Apache Airflow variables to be exposed for the parameters
    variables_params: [env]
  
    # Optional resources specification
    resources:
        # Default configuration used by all nodes that do not declare the
        # resource configuration. It's optional. If node does not declare the resource
        # configuration, __default__ is assigned by default, otherwise cluster defaults
        # will be used.
        __default__:
            # Optional labels to be put into pod node selector
            node_selectors:
              #Labels are user provided key value pairs
              node_pool_label/k8s.io: example_value
            # Optional labels to apply on pods
            labels:
              running: airflow
            # Optional annotations to apply on pods
            annotations:
              iam.amazonaws.com/role: airflow
            # Optional list of kubernetes tolerations
            tolerations:
                - key: "group"
                  value: "data-processing"
                  effect: "NoExecute"
                - key: "group"
                  operator: "Equal",
                  value: "data-processing",
                  effect: "NoSchedule"
            requests:
                #Optional amount of cpu resources requested from k8s
                cpu: "1"
                Optional amount of memory resource requested from k8s
                memory: "1Gi"
            limits:
                #Optional amount of cpu resources limit on k8s
                cpu: "1"
                #Optional amount of memory resource limit on k8s
                memory: "1Gi"
        # Other arbitrary configurations to use, for example to indicate some exception resources
        huge_machines:
            node_selectors:
                big_node_pool: huge.10x
            requests:
                cpu: "16"
                memory: "128Gi"
            limits:
                cpu: "32"
                memory: "256Gi"
    # Optional external dependencies configuration
    external_dependencies:
        # Can just select dag as a whole 
        - dag_id: upstream-dag
        # or detailed
        - dag_id: another-upstream-dag
        # with specific task to wait on
          task_id: with-precise-task
        # Maximum time (minute) to wait for the external dag to finish before this
        # pipeline fails, the default is 1440 == 1 day  
          timeout: 2
        # Checks if the external dag exists before waiting for it to finish. If it
        # does not exists, fail this pipeline. By default is set to true. 
          check_existence: False
        # Time difference with the previous execution to look at (minutes),
        # the default is 0 meaning no difference
          execution_delta: 10

    # Optional authentication to MLflow API    
    authentication:
      # Strategy that generates the tokens, supported values are: 
      # - Null
      # - GoogleOAuth2 (generating OAuth2 tokens for service account provided by GOOGLE_APPLICATION_CREDENTIALS)
      # - Vars (credentials fetched from airflow Variable.get - specify variable keys,
      # matching MLflow authentication env variable names, in `params`,
      # e.g. ["MLFLOW_TRACKING_USERNAME", "MLFLOW_TRACKING_PASSWORD"])
      type: GoogleOAuth2 
      #params: []

    # Optional custom kubermentes pod templates applied on nodes basis
    kubernetes_pod_templates:
    # Name of the node you want to apply the custom template to.
    # if you specify __default__, this template will be applied to all nodes.
    # Otherwise it will be only applied to nodes tagged with `k8s_template:<node_name>`
      spark:
    # Kubernetes pod template.
    # It's the full content of the pod-template file (as a string)
    # `run_config.volume` and `MLFLOW_RUN_ID` env are disabled when this is set.
    # Note: python F-string formatting is applied to this string, so
    # you can also use some dynamic values, e.g. to calculate pod name.
        template: |-
          type: Pod
          metadata:
            name: {PodGenerator.make_unique_pod_id('{{ task_instance.task_id }}')}
            labels:
              spark_driver: {'{{ task_instance.task_id }}'}

    # Optionally, you can also override the image
    #   image:
    
    # Optional spark configuration
    spark:
      # Type of spark clusters to use, supported values: dataproc, k8s, kubernetes, custom
      type: dataproc
      # Optional factory of spark operators class
      operator_factory: my_project.factories.OperatorFactory
      # Region indicates location of cluster for public cloud configurations, for example region in GCP
      region: europe-west1
      # Project indicates logical placement inside public cloud configuration, for example project in GCP
      project_id: target-project
      # Name of the cluster to be created 
      cluster_name: ephemeral
      # Location where the spark artifacts are uploaded
      artifacts_path: gs://dataproc-staging-europe-west2-546213781-jabcdefp4/packages
      # Optional path in the project to the script portion preprended to generated init script
      user_init_path: relative_location_to_src/init_script.sh
      # Optional path in the project to the script portion appended to generated init script
      user_post_init_path: relative_location_to_src/post_init_script.sh
      # Optional configuration of the cluster, used during cluster creation, depends on type of the cluster
      cluster_config: # example dataproc configuration
        master_config:
          disk_config:
            boot_disk_size_gb: 35
        worker_config:
          disk_config:
            boot_disk_size_gb: 35
```

## Indicate resources in pipeline nodes

Every node declared in `kedro` pipelines is executed inside pod. Pod definition declares resources to be used based
on provided plugin configuration and presence of the tag `resources` in `kedro` node definition.

If no such tag is present, plugin will assign `__default__` from plugin `resources` configuration.

CHANGELOG.md
# Changelog

## [Unreleased]
-   Add e2e tests with Cloud Composer 2

## [0.8.0] - 2022-03-23

-   Support spark projects on K8S
-   FIX: handle `KEDRO_ENV` environment variable
-   FIX: Removed hardcoded project path from spark job

## [0.7.3] - 2021-11-16

-   Take DAG status from the final task
-   Fix finding pyspark DataFrames
-   `MLFLOW_RUN_ID` passed as environment variable to dataproc oriented pipelines

## [0.7.2] - 2021-10-25

-   Support annotations with quotes
-   For pyspark projects, allows to configure post script for dataproc initialization
-   FIX: Broken support for kedro&lt;0.17

## [0.7.1] - 2021-10-21

-   Support for failure notifications via slack
-   FIX: Missing jinja template for dataproc init script

## [0.7.0] - 2021-10-19

-   Schedule supports `dag-name` parameter
-   Support for kedro with pyspark, using Google Dataproc
-   Support for custom pod templates
-   FIX: adding missing dependency package: tabulate
-   FIX: Fix default config template
-   Support populating k8s node env_vars from Airflow variables
-   Generalize auth handler
-   Added `VarsAuthHandler` for MLflow authentication which gets credentials from Airflow variables
-   Changed logging level for pod creation request to debug

## [0.6.7] - 2021-09-01

-   Support for generation of authentication header for secured MLflow API endpoint (via GOOGLE_APPLICATION_CREDENTIALS)

## [0.6.6] - 2021-08-16

-   Support for passing Authorization header for secured Airflow API endpoint (via env variable: `AIRFLOW_API_TOKEN`)
-   Logging `dag_id` and `execution_date` in mlflow run params

## [0.6.5] - 2021-08-05

-   FIX: Adjust service account setup for image based tasks

## [0.6.4] - 2021-08-05

-   FIX: Adjusted operators to make them compatible with Airflow >= 2.1.1
-   FIX: Restore dependency versions that release process bumped unintentionally

## [0.6.3] - 2021-08-04

-   FIX: Avoid file based Jinja template for `data-volume-init`

## [0.6.2] - 2021-08-04

-   FIX: Add missing Jinja template for `data-volume-init` to module manifest 

## [0.6.1] - 2021-08-04

-   FIX: service_account_name or image_pull_secrets should be passed to `data-volume-init` step as well

## [0.6.0] - 2021-07-29

-   Added option to specify service_account_name or image_pull_secrets to executed dag. 

## [0.5.4] - 2021-07-21

-   Run config contains optional start_date in format YYYYMMDD, if not specified default is left to `days_ago(2)` 

## [0.5.3] - 2021-07-12

-   Support for airflow macro parameters and variables

## [0.5.2] - 2021-07-05

-   Support for Secrets in k8s
-   FIX: DeletePipelineStorageOperator was missing trigger rule 'all_done'

## [0.5.1] - 2021-05-17

-   Docker image added as a Mlflow run parameter (to support kedro inference pipeline)

## [0.5.0] - 2021-04-30

-   External dependencies can be added as optionals in configuration
-   Support for labels, tolerations and annotations in k8s
-   Logging added to operators

## [0.4.0] - 2021-04-20

-   Support of S3 as DAG destination
-   Operators extracted from DAG template as reusable components
-   Selection of specific pipeline by name from CLI

## [0.3.0] - 2021-04-15

-   Resources configuration added that allows describing cpu and memory resources required in k8s by pods
-   Shared persistent volume can be made optional
-   Pod startup timeout is configurable, with default to 600 seconds

## [0.2.0] - 2021-04-01

-   Command `list-pipelines` allows display of pipelines which were generated with the plugin
-   Command `ui` opens webbrowser with Apache Airflow
-   Move options to config file and restructure
-   `MLflow` enabled only if `kedro-mlflow` present in dependencies and configuration is in place 
-   Command `init` to initialize configuration for the plugin in kedro project

## [0.1.2] - 2021-03-24

-   Allow override `image` parameter from CLI for `upload_pipeline`, `run_once` and `compile`
-   Conditional SSL verification for Apache Airflow client, via AIRFLOW**CLIENT**SSL_VERIFY environment variable
-   `run-once` with `wait-for-completion` checks for task instance failure as well

## [0.1.1] - 2021-03-19

-   Inits temporary volume with data from image kedro data directory (/home/kedro/data)
-   Increased startup time for pods to 10 minutes

## [0.1.0] - 2021-03-17

-   Creates mlflow experiment on pipeline start if it does not exist
-   Temporary persistent volumes generation and removal for pipelines
-   `upload-pipeline` command that generates DAG to Airflow directory 
-   `schedule` command that generates DAG to Airflow directory with the given schedule
-   `run-once` command that uploads pipeline to Airflow and creates the DAG run

## [0.0.4] - 2021-03-10

### Fixed

-   Added DAG template to package manifest

## [0.0.3] - 2021-03-09

### Fixed

-   Drop dependency on `airflow`

## [0.0.2] - 2021-03-09

### Added

-   Initial implementation of `kedro_airflow_k8s` plugin stub. 

[Unreleased]: https://github.com/getindata/kedro-airflow-k8s/compare/0.8.0...HEAD


kedro_airflow_k8s/config.py
import os

from kedro.config import MissingConfigException

DEFAULT_CONFIG_TEMPLATE = """
# Base url of the Apache Airflow, should include the schema (http/https)
host: {url}

# Directory from where Apache Airflow is reading DAGs definitions
output: {output}

# Type of execution engine to use.
#
# For **Kubernetes** deployment and in-cluster Airflow, the section
# looks like:
# execution_engine:
#   type: kubernetes
#
# For **AWS Fargate** deployment, it contains all the information required
# to run the containers in serverless manner:
# execution_engine:
#   type: aws-fargate
#   cluster: ml-pipelines
#   task_definition: my-first-ml-pipeline
#   subnets:
#   - subnet-001122334455
#   security_groups:
#   - sg-0248abcd0248
execution_engine:
  type: kubernetes

# Configuration used to run the pipeline
run_config:

    # Name of the image to run as the pipeline steps
    image: {image}

    # Pull policy to be used for the steps. Use Always if you push the images
    # on the same tag, or Never if you use only local images
    image_pull_policy: IfNotPresent

    # Pod startup timeout in seconds
    startup_timeout: 600

    # Namespace for Airflow pods to be created
    namespace: airflow

    # Name of the Airflow experiment to be created
    experiment_name: {project}

    # Name of the dag as it's presented in Airflow
    run_name: {run_name}

    # Apache Airflow cron expression for scheduled runs
    cron_expression: "@daily"

    # Optional start date in format YYYYMMDD
    #start_date: "20210721"

    # Optional pipeline description
    #description: "Very Important Pipeline"

    # Comma separated list of image pull secret names
    #image_pull_secrets: my-registry-credentials

    # Service account name to execute nodes with
    #service_account_name: default

    # List of handlers executed after task failure
    failure_handlers: []

    # Optional volume specification
    volume:
        # Storage class - use null (or no value) to use the default storage
        # class deployed on the Kubernetes cluster
        storageclass: # default
        # The size of the volume that is created. Applicable for some storage
        # classes
        size: 1Gi
        # Access mode of the volume used to exchange data. ReadWriteMany is
        # preferred, but it is not supported on some environements (like GKE)
        # Default value: ReadWriteOnce
        #access_modes: [ReadWriteMany]
        # Flag indicating if the data-volume-init step (copying raw data to the
        # fresh volume) should be skipped
        skip_init: False
        # Allows to specify fsGroup executing pipelines within containers
        # Default: root user group (to avoid issues with volumes in GKE)
        owner: 0
        # Tells if volume should not be used at all, false by default
        disabled: False

    # List of optional secrets specification
    secrets:
            # deploy_type: The type of secret deploy in Kubernetes, either `env` or
            # `volume`
        -   deploy_type: "env"
            # deploy_target: (Optional) The environment variable when `deploy_type` `env`
            # or file path when `deploy_type` `volume` where expose secret. If `key` is
            # not provided deploy target should be None.
            deploy_target: "SQL_CONN"
            # secret: Name of the secrets object in Kubernetes
            secret: "airflow-secrets"
            # key: (Optional) Key of the secret within the Kubernetes Secret if not
            # provided in `deploy_type` `env` it will mount all secrets in object
            key: "sql_alchemy_conn"

    # Apache Airflow macros to be exposed for the parameters
    # List of macros can be found here:
    # https://airflow.apache.org/docs/apache-airflow/stable/macros-ref.html
    macro_params: [ds, prev_ds]

    # Apache Airflow variables to be exposed for the parameters
    variables_params: [env]

    # Optional resources specification
    #resources:
        # Default configuration used by all nodes that do not declare the
        # resource configuration. It's optional. If node does not declare the resource
        # configuration, __default__ is assigned by default, otherwise cluster defaults
        # will be used.
        #__default__:
            # Optional labels to be put into pod node selector
            #node_selectors:
              #Labels are user provided key value pairs
              #node_pool_label/k8s.io: example_value
            # Optional labels to apply on pods
            #labels:
              #running: airflow
            # Optional annotations to apply on pods
            #annotations:
              #iam.amazonaws.com/role: airflow
              #vault.hashicorp.com/agent-inject-template-foo: |
              # {{- with secret "database/creds/db-app" -}}
              # postgres://{{ .Data.username }}:{{ .Data.password }}@postgres:5432/mydb
              # {{- end }}
            # Optional list of kubernetes tolerations
            #tolerations:
                #- key: "group"
                  #value: "data-processing"
                  #effect: "NoExecute"
                #- key: "group"
                  #operator: "Equal",
                  #value: "data-processing",
                  #effect: "NoSchedule"
            #requests:
                #Optional amount of cpu resources requested from k8s
                #cpu: "1"
                #Optional amount of memory resource requested from k8s
                #memory: "1Gi"
            #limits:
                #Optional amount of cpu resources limit on k8s
                #cpu: "1"
                #Optional amount of memory resource limit on k8s
                #memory: "1Gi"
        # Other arbitrary configurations to use
        #custom_resource_config_name:
            # Optional labels to be put into pod node selector
            #labels:
                #Labels are user provided key value pairs
                #label_key: label_value
            #requests:
                #Optional amount of cpu resources requested from k8s
                #cpu: "1"
                #Optional amount of memory resource requested from k8s
                #memory: "1Gi"
            #limits:
                #Optional amount of cpu resources limit on k8s
                #cpu: "1"
                #Optional amount of memory resource limit on k8s
                #memory: "1Gi"

    # Optional external dependencies configuration
    #external_dependencies:
        # Can just select dag as a whole
        #- dag_id: upstream-dag
        # or detailed
        #- dag_id: another-upstream-dag
        # with specific task to wait on
        #  task_id: with-precise-task
        # Maximum time (minute) to wait for the external dag to finish before this
        # pipeline fails, the default is 1440 == 1 day
        #  timeout: 2
        # Checks if the external dag exists before waiting for it to finish. If it
        # does not exists, fail this pipeline. By default is set to true.
        #  check_existence: False
        # Time difference with the previous execution to look at (minutes),
        # the default is 0 meaning no difference
        #  execution_delta: 10
    # Optional authentication to MLflow API
    #authentication:
      # Strategy that generates the credentials, supported values are:
      # - Null
      # - GoogleOAuth2 (generating OAuth2 tokens for service account provided by
      # GOOGLE_APPLICATION_CREDENTIALS)
      # - Vars (credentials fetched from airflow Variable.get - specify variable keys,
      # matching MLflow authentication env variable names, in `params`,
      # e.g. ["MLFLOW_TRACKING_USERNAME", "MLFLOW_TRACKING_PASSWORD"])
      #type: GoogleOAuth2
      #params: []
    #spark:
    #  submit_job_operator:
    # Airflow operator to use for submitting Spark job: SparkSubmitOperator,
    # DataprocSubmitJobOperator or KubernetesSparkOperator
    #  region: None
    #  project_id: None
    #  cluster_name: None
    #  create_cluster: False

    # Optional custom kubermentes pod templates applied on nodes basis
    #kubernetes_pod_templates:
    # Name of the node you want to apply the custom template to.
    # if you specify __default__, this template will be applied to all nodes.
    # Otherwise it will be only applied to nodes tagged with `k8s_template:<node_name>`
    #  node_name:

    # Kubernetes pod template.
    # It's the full content of the pod-template file (as a string)
    # `run_config.volume` and `MLFLOW_RUN_ID` env are disabled when this is set.
    # Note: python F-string formatting is applied to this string, so
    # you can also use some dynamic values, e.g. to calculate pod name.
    #    template:

    # Optionally, you can also override the image
    #    image:
    # ____ EXAMPLE _______________
    #
    #kubernetes_pod_templates:
    #  spark:
    #    template: |-
    #      apiVersion: v1
    #      kind: Pod
    #      metadata:
    #        name: newname
    #       spec:
    #         containers:
    #           - name: base
    #         env:
    #           - name: CUSTOM_ENV
    #             value: env1
    #

    # Configuration for spark jobs
    # spark:
    # type: k8s
    # cluster_name: spark_k8s
    #   run_script: local:///path/to/run/script/in/image.py
"""


class Config(object):
    def __init__(self, raw):
        self._raw = raw

    def _get_or_default(self, prop, default):
        return self._raw.get(prop, default)

    def _get_or_fail(self, prop):
        if prop in self._raw.keys():
            return self._raw[prop]
        else:
            raise MissingConfigException(
                f"Missing required configuration: '{self._get_prefix()}{prop}'."
            )

    def _get_prefix(self):
        return ""

    def __eq__(self, other):
        return self._raw == other._raw


class ResourceNodeConfig(Config):
    @property
    def cpu(self):
        return self._get_or_default("cpu", None)

    @property
    def memory(self):
        return self._get_or_default("memory", None)


class ResourceConfig(Config):
    @property
    def annotations(self):
        return self._get_or_default("annotations", {})

    @property
    def tolerations(self):
        return self._get_or_default("tolerations", {})

    @property
    def node_selectors(self):
        return self._get_or_default("node_selectors", {})

    @property
    def labels(self):
        return self._get_or_default("labels", {})

    @property
    def requests(self):
        return ResourceNodeConfig(self._get_or_default("requests", {}))

    @property
    def limits(self):
        return ResourceNodeConfig(self._get_or_default("limits", {}))


class ResourcesConfig(Config):
    def __getattr__(self, item):
        return self[item]

    def __getitem__(self, item):
        return ResourceConfig(self._get_or_default(item, {}))


class ExternalDependencyConfig(Config):
    @property
    def dag_id(self):
        return self._get_or_fail("dag_id")

    @property
    def task_id(self):
        return self._get_or_default("task_id", None)

    @property
    def check_existence(self):
        return self._get_or_default("check_existence", True)

    @property
    def execution_delta(self):
        return self._get_or_default("execution_delta", 0)

    @property
    def timeout(self):
        return self._get_or_default("timeout", 60 * 24)


class AuthenticationConfig(Config):
    @property
    def type(self):
        return self._get_or_default("type", "Null")

    @property
    def params(self):
        return self._get_or_default("params", [])


class SparkConfig(Config):
    @property
    def type(self):
        return self._get_or_default("type", "none")

    @property
    def region(self):
        return self._get_or_default("region", "None")

    @property
    def cluster_name(self):
        return self._get_or_default("cluster_name", "None")

    @property
    def project_id(self):
        return self._get_or_default("project_id", "None")

    @property
    def operator_factory(self):
        return self._get_or_default("operator_factory", None)

    @property
    def artifacts_path(self):
        return self._get_or_default("artifacts_path", None)

    @property
    def user_init_path(self):
        return self._get_or_default("user_init_path", None)

    @property
    def user_post_init_path(self):
        return self._get_or_default("user_post_init_path", None)

    @property
    def cluster_config(self):
        data = self._get_or_default("cluster_config", {})
        if self.type in ["k8s", "kubernetes"]:
            return SparkK8SConfig(data)
        return data

    @property
    def requires_artifacts_dump(self):
        return self.type not in ["k8s", "kubernetes"]


class SparkK8SStorageConfig(Config):
    @property
    def class_name(self):
        return self._get_or_default("class_name", "standard")

    @property
    def size(self):
        return self._get_or_default("size", None)


class SparkK8SConfig(Config):
    @property
    def run_script(self):
        return self._get_or_fail("run_script")

    @property
    def image(self):
        return self._get_or_default("image", None)

    @property
    def conf(self):
        return self._get_or_default("conf", {})

    @property
    def driver_port(self):
        return self._get_or_default("driver_port", None)

    @property
    def block_manager_port(self):
        return self._get_or_default("block_manager_port", None)

    @property
    def secrets(self):
        return self._get_or_default("secrets", {})

    @property
    def labels(self):
        return self._get_or_default("labels", {})

    @property
    def local_storage(self):
        return SparkK8SStorageConfig(self._get_or_default("local_storage", {}))

    @property
    def env_vars(self):
        return self._get_or_default("env_vars", {})

    @property
    def requests(self):
        return ResourceNodeConfig(self._get_or_default("requests", {}))

    @property
    def limits(self):
        return ResourceNodeConfig(self._get_or_default("limits", {}))

    @property
    def num_executors(self):
        return self._get_or_default("num_executors", "1")

    @property
    def jars(self):
        return self._get_or_default("jars", None)

    @property
    def repositories(self):
        return self._get_or_default("repositories", None)

    @property
    def packages(self):
        return self._get_or_default("packages", None)


class RunConfig(Config):
    @property
    def image(self):
        return self._get_or_fail("image")

    @property
    def image_pull_policy(self):
        return self._get_or_default("image_pull_policy", "IfNotPresent")

    @property
    def startup_timeout(self):
        return self._get_or_default("startup_timeout", 600)

    @property
    def namespace(self):
        return self._get_or_fail("namespace")

    @property
    def failure_handlers(self):
        cfg = self._get_or_default("failure_handlers", [])
        supported_types = FailureHandlerConfig.supported_types()
        return [
            FailureHandlerConfig(handler)
            for handler in cfg
            if handler["type"] in supported_types
        ]

    @property
    def experiment_name(self):
        return self._get_or_fail("experiment_name")

    @property
    def run_name(self):
        return self._get_or_default("run_name", self.experiment_name)

    @property
    def cron_expression(self):
        return self._get_or_default("cron_expression", "@daily")

    @property
    def start_date(self):
        start_date = self._get_or_default("start_date", None)
        if start_date:
            start_date = str(start_date)
        return start_date

    @property
    def description(self):
        return self._get_or_default("description", None)

    @property
    def volume(self):
        cfg = self._get_or_default("volume", {})
        return VolumeConfig(cfg)

    @property
    def secrets(self):
        cfg = self._get_or_default("secrets", [])
        return [SecretConfig(secret) for secret in cfg]

    @property
    def macro_params(self):
        return self._get_or_default("macro_params", [])

    @property
    def variables_params(self):
        return self._get_or_default("variables_params", [])

    @property
    def resources(self):
        cfg = self._get_or_default("resources", {})
        return ResourcesConfig(cfg)

    @property
    def external_dependencies(self):
        deps = self._get_or_default("external_dependencies", [])
        return [ExternalDependencyConfig(cfg) for cfg in deps]

    @property
    def image_pull_secrets(self):
        return self._get_or_default("image_pull_secrets", None)

    @property
    def service_account_name(self):
        return self._get_or_default("service_account_name", None)

    @property
    def auth_config(self):
        cfg = self._get_or_default(
            "authentication", {"type": "Null", "params": []}
        )
        return AuthenticationConfig(cfg)

    @property
    def spark(self):
        cfg = self._get_or_default("spark", {})
        return SparkConfig(cfg)

    @property
    def env_vars(self):
        return self._get_or_default("env_vars", [])

    @property
    def kubernetes_pod_templates(self):
        cfg = self._get_or_default("kubernetes_pod_templates", {})
        return KubernetesPodTemplates(cfg)

    def _get_prefix(self):
        return "run_config."


class VolumeConfig(Config):
    @property
    def disabled(self):
        return self._get_or_default("disabled", False)

    @property
    def storageclass(self):
        return self._get_or_default("storageclass", None)

    @property
    def size(self):
        return self._get_or_default("size", "1Gi")

    @property
    def access_modes(self):
        return self._get_or_default("access_modes", ["ReadWriteOnce"])

    @property
    def skip_init(self):
        return self._get_or_default("skip_init", False)

    @property
    def owner(self):
        return self._get_or_default("owner", 0)

    def _get_prefix(self):
        return "run_config.volume."


class FailureHandlerConfig(Config):
    @staticmethod
    def supported_types():
        return ["slack"]

    @property
    def type(self):
        return self._get_or_fail("type")

    @property
    def connection_id(self):
        return self._get_or_fail("connection_id")

    @property
    def message_template(self):
        return self._get_or_default("message_template", "Task failed!")


class ExecutionEngine(Config):
    @property
    def type(self):
        return self._get_or_fail("type")

    @property
    def params(self):
        return self._raw


class SecretConfig(Config):
    @property
    def deploy_type(self):
        return self._get_or_default("deploy_type", "env")

    @property
    def deploy_target(self):
        return self._get_or_default("deploy_target", None)

    @property
    def secret(self):
        return self._get_or_fail("secret")

    @property
    def key(self):
        return self._get_or_default("key", None)

    def _get_prefix(self):
        return "run_config.secrets."


class PluginConfig(Config):
    @property
    def host(self):
        return self._get_or_fail("host")

    @property
    def output(self):
        return self._get_or_fail("output")

    @property
    def run_config(self):
        cfg = self._get_or_default("run_config", {})
        return RunConfig(cfg)

    @property
    def execution_engine(self):
        cfg = self._get_or_default("execution_engine", {"type": "kubernetes"})
        return ExecutionEngine(cfg)

    @staticmethod
    def sample_config(**kwargs):
        return DEFAULT_CONFIG_TEMPLATE.format(**kwargs)

    @staticmethod
    def initialize_github_actions(project_name, where, templates_dir):
        os.makedirs(where / ".github/workflows", exist_ok=True)
        for template in ["on-merge-to-master.yml", "on-push.yml"]:
            file_path = where / ".github/workflows" / template
            template_file = templates_dir / f"github-{template}"
            with open(template_file, "r") as tfile, open(file_path, "w") as f:
                f.write(tfile.read().format(project_name=project_name))


class KubernetesPodTemplate(Config):
    @property
    def template(self):
        return self._get_or_default("template", None)

    @property
    def image(self):
        return self._get_or_default("image", None)

    def __len__(self):
        return len(self._raw)


class KubernetesPodTemplates(Config):
    def __getattr__(self, item):
        return self[item]

    def __getitem__(self, item):
        return KubernetesPodTemplate(self._get_or_default(item, {}))

    def __len__(self):
        return len(self._raw)
